File: typescript-examples/computer-use-automation/lib/anthropic/tools/browser.ts
import type { Page } from 'playwright';
import { ToolError } from '../types/computer';
import { goToUrl } from '@intuned/browser';

export interface BrowserToolResult {
  output?: string;
  error?: string;
  base64Image?: string;
}

export class BrowserTool {
  name: 'browser' = 'browser';
  protected page: Page;
  protected _screenshotDelay = 2.0;

  constructor(page: Page) {
    this.page = page;
  }

  toParams(): any {
    return {
      name: this.name,
      type: 'custom',
      description: 'Browser navigation tool for controlling the web browser',
      input_schema: {
        type: 'object',
        properties: {
          action: {
            type: 'string',
            enum: ['go_to_url', 'go_back', 'go_forward', 'refresh'],
            description: 'The navigation action to perform',
          },
          url: {
            type: 'string',
            description: 'The URL to navigate to (required for go_to_url action)',
          },
        },
        required: ['action'],
      },
    };
  }

  async screenshot(): Promise<BrowserToolResult> {
    try {
      console.log('Browser tool: taking screenshot...');
      await new Promise(resolve => setTimeout(resolve, this._screenshotDelay * 1000));
      const screenshot = await this.page.screenshot({ type: 'png' });
      console.log('Browser tool: screenshot taken, size:', screenshot.length, 'bytes');
      return {
        base64Image: screenshot.toString('base64'),
      };
    } catch (error) {
      throw new ToolError(`Failed to take screenshot: ${error}`);
    }
  }

  async call(params: { action: string; url?: string }): Promise<BrowserToolResult> {
    const { action, url } = params;

    switch (action) {
      case 'go_to_url':
        if (!url) {
          throw new ToolError('url is required for go_to_url action');
        }
        console.log(`Navigating to URL: ${url}`);
        await goToUrl({ page: this.page, url });
        await this.page.waitForTimeout(2000);
        console.log(`Successfully navigated to: ${url}`);
        return await this.screenshot();

      case 'go_back':
        console.log('Navigating back');
        try {
          await this.page.goBack({ waitUntil: 'domcontentloaded', timeout: 30000 });
        } catch (error) {
          return {
            output: 'Cannot go back - no previous page in history',
          };
        }
        await this.page.waitForTimeout(1000);
        return await this.screenshot();

      case 'go_forward':
        console.log('Navigating forward');
        try {
          await this.page.goForward({ waitUntil: 'domcontentloaded', timeout: 30000 });
        } catch (error) {
          return {
            output: 'Cannot go forward - no forward page in history',
          };
        }
        await this.page.waitForTimeout(1000);
        return await this.screenshot();

      case 'refresh':
        console.log('Refreshing page');
        await this.page.reload({ waitUntil: 'domcontentloaded', timeout: 30000 });
        await this.page.waitForTimeout(1000);
        return await this.screenshot();

      default:
        throw new ToolError(`Unknown browser action: ${action}`);
    }
  }
}

File: python-examples/computer-use-automation/lib/anthropic/tools/browser.py
"""
Browser navigation tool using Playwright.
"""

import base64
from typing import Literal
from playwright.async_api import Page
from intuned_browser import go_to_url

from anthropic.types.beta import BetaToolUnionParam

from .base import BaseAnthropicTool, ToolError, ToolResult


class BrowserTool(BaseAnthropicTool):
    """
    A tool that allows the agent to navigate to URLs using the browser.
    """

    name: Literal["browser"] = "browser"
    api_type: Literal["custom"] = "custom"
    page: Page | None = None

    def __init__(self, page: Page | None = None):
        super().__init__()
        self.page = page

    def to_params(self) -> BetaToolUnionParam:
        return {
            "type": "custom",
            "name": self.name,
            "description": "Navigate to a URL in the browser. Use this tool to go to websites directly instead of manually typing URLs.",
            "input_schema": {
                "type": "object",
                "properties": {
                    "url": {
                        "type": "string",
                        "description": "The URL to navigate to. Include the protocol (http:// or https://) if needed, otherwise it will be added automatically.",
                    }
                },
                "required": ["url"],
            },
        }

    async def __call__(
        self,
        *,
        url: str,
        **kwargs,
    ) -> ToolResult:
        if not self.page:
            raise ToolError("Playwright page not initialized")

        if not url:
            raise ToolError("URL is required")

        # Add https:// if no protocol is specified
        if not url.startswith(("http://", "https://")):
            url = f"https://{url}"

        try:
            # Navigate to the URL
            await go_to_url(page=self.page, url=url)
            
            # Wait a bit for the page to render
            await self.page.wait_for_timeout(2000)
            response = None  # go_to_url doesn't return a response object
            
            # Take a screenshot
            screenshot_bytes = await self.page.screenshot(type="png")
            
            status_text = "success" if response and response.ok else "failed"
            output = f"Navigated to {url} - Status: {status_text}"
            
            return ToolResult(
                output=output,
                base64_image=base64.b64encode(screenshot_bytes).decode()
            )
        except Exception as e:
            raise ToolError(f"Failed to navigate to {url}: {str(e)}")

